intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/staticanalysis/IntelliJMarkupModelListener.java

```diff
@@ -0,0 +1,220 @@
+
+                        // We batch up changes and only transfer them on every save. This is in-line with the Eclipse
+                        // interface, which only exposes listeners for `POST_BUILD`. Therefore, cache all warnings in
+                        // {@link IntelliJMarkupModelListener#generatedWarnings} and {@link IntelliJMarkupModelListener#warnings}
+                        // and flush these warnings after the fact.
+                        documentMessageBusConnection.subscribe(AppTopics.FILE_DOCUMENT_SYNC, new FileDocumentManagerAdapter() {
+                            @Override
+                            public void beforeDocumentSaving(@NotNull com.intellij.openapi.editor.Document savedDocument) {
+                                if (intellijDocument.equals(savedDocument)) {
+                                    markupModelListener.flushForDocument();
+                                }
+                            }
+                        });
+                        codeAnalyzerMessageBusConnection.disconnect();
+                    }
+                }
+            });
+        });
+
+        return markupModelListener;
+    }
+
+    private void processWarningSnapshot(RangeHighlighter[] highlighters) {
+        if (highlighters.length == 0) {
+            return;
+        }
+
+        // markupModel.getAllHighlighters returns a list of RangeHighlighter,
+        // even though we know they are all instances of RangeHighlighterEx.
+        // Therefore do an instanceof check and then explicitly cast them
+        // to use them in the other methods
+        final Stream<RangeHighlighterEx> rangeHighlighters = Arrays.stream(highlighters)
+                .filter(RangeHighlighterEx.class::isInstance)
+                .map(RangeHighlighterEx.class::cast);
+
+        final List<Warning<String>> warnings = rangeHighlighters
+                .filter(IntelliJMarkupModelListener::isWarningRangeHighlighter)
+                .map(rangeHighlighter -> createWarningFromRangeHighlighter(rangeHighlighter, null))
+                .map(IntelliJMarkupModelListener::classifyWarning)
+                .collect(Collectors.toList());
+
+        this.trackingEventManager.addEvent(new FileWarningSnapshotEvent(this.document.prepareDocument(), warnings));
+    }
+
+    private void flushForDocument() {
+        addCreatedWarnings(this.trackingEventManager, this.generatedWarnings.stream().map(IntelliJMarkupModelListener::classifyWarning), this.document);
+        this.generatedWarnings.clear();
+
+        addRemovedWarnings(this.trackingEventManager, this.warnings.stream().map(IntelliJMarkupModelListener::classifyWarning), this.document);
+        this.warnings.clear();
+    }
+
+    @Override
+    public void afterAdded(@NotNull RangeHighlighterEx rangeHighlighterEx) {
+        if (isWarningRangeHighlighter(rangeHighlighterEx)) {
+            final DateTime creationTime = DateTime.now();
```

Comment:
does this really need to be final?

Reply:
They are never changed nor should they, so this is mostly a hint to the compiler for optimization purposes.

Reply:
I would prefer "final" to be removed if it is not absolutely necessary to compile. It burdens people's mind as to why it can't just be a normal local variable.